src/mpi/coll/allreduce/allreduce_intra_recursive_multiplying.c

```diff
@@ -0,0 +1,230 @@
+/*
+ * Copyright (C) by Argonne National Laboratory
+ *     See COPYRIGHT in top-level directory
+ */
+
+#include "mpiimpl.h"
+
+#include <math.h>
+
+/*
+ * Algorithm: Recursive Multiplying
+ *
+ * This algorithm is a generalization of the recursive doubling algorithm,
+ * and it has three stages. In the first stage, ranks swap data in groups
+ * of k nodes. Then, the "leaders" of each group (the first rank) exchange
+ * data with all other group leaders in rounds with increasing distance (k,
+ * k^2, ...). Lastly, the leaders disperse the inter-group information 
+ * back to their group. Setting k according to the network hierarchy 
+ * (e.g., the number of NICs in a node) can improve performance.
```

Comment:
Thanks for the explanation! It helps.

There are 3 concepts here need individually convey 
1. A group of k processes do one round of *alltoall* communication followed with local reductions
2. Recursively reforming the groups with spacing of 1, k, k^2 ...
3. The first round and last round are special. Are you collapsing extra processes to pofk?